### DISCUSSION TITLE
helper to create nuxi custom command with nuxt context

### DISCUSSION THREAD

**Discussion by stafyniaksacha** [nuxt/nuxt#17195] [Ideas]
Now we have a proper server side environment set up via nitro, it would be great to have helpers to create CLI/Cron scripts with same configuration context as our server controllers.

The idea would be to have a new autoloaded directory (ex `./scripts/[script-name].ts`) where we can register commands like:

```ts
// ./scripts/hello.ts

// auto imported nuxi helpers
import { parseArgs, defineArgs, printHelp } from 'nuxi'

export default defineNuxiCommand(async () => {
  // define script arguments
  const args = defineArgs({
    from: {
      type: String,
      required: true,
    }
  })

  const { from } = parseArgs(args)

  if (!from) {
    printHelp(args)
    return
  }

  // have access to useRuntimeConfig
  const config = useRuntimeConfig()

  // have access to nuxtApp and everything provided by modules
  const { $sentry, $pinia } = useNuxtApp()

  // have access to nitro context
  await useStorage().getItem('test:foo')
})
```

In development, we could do:
- `nuxi run --list` to view available commands
- `nuxi run [script-name]`

When building, those scripts should be embedded in the server's output, so we can use them with
- `node .output/server/scripts/[script-name].mjs`

**Comment by pi0** [comment]
Hi @stafyniaksacha and thanks for sharing the idea.

A short answer: With Nuxt 3 we have two explicitly separated contexts:
- Build Context: Nuxt core, vite, kit, nuxi cli, and modules are in this namespace
- Runtime Context: Vue app, Nitro server and plugins are in this context

This allows portability, and all new benefits nitro introduced for production deployments. Therefore by design, it is impossible that a nuxi cli command in the build context to access runtime nor nuxi is intended to be used in production (but only development and build)


🇦 If you need a portable task script that also runs in production by manual trigger and has access to the runtime, you can create an API route in `server/routes` or `server/api`. It can access storage and runtime config.

🇧 If you need a background task runner or one time startup task script, you can use a nitro plugin by placing in `server/plugins`.

🇨 If you need a nuxi CLI script for development and build time (not production), you can create it in `bin` directory and run it with `jiti bin/script.ts`. We plan to make it easier with https://github.com/nuxt/framework/issues/4060 to build a custom nuxi script.


What I think we can improve in Nuxt according to this idea:

- Ship support of external nuxi commands
- Documented programmatic usage of `setupNuxt` that allows usage of `@nuxt/kit` utils or a shortcut that sets up Nuxt without build to access Nuxt (build context)
- Introduce some identical utilities from `@nuxt/kit` such as `useRuntimeConfig` and `useStorage` (already exposed from nitro in build context). Maybe even `$fetch` but it is only for prerender phase.

Again please note that this improvement has limitations for nuxi commands:

- Runtime is in no way accessible
- They are suitable for purpose of development and build scripts
- A limited set of API would be available from `@nuxt/kit` and new nuxt modules can contribute to add more.
- Need `package.json` and `node_modules` dependencies

In contrast, alternative nitro plugins:

- Have full access to runtime and runtime composables
- Are best option for production but also can be used for development and build (within prerender phase)
- Do not need `node_modules` and are fully standalone 
- Can be triggered on server startup or on demand by an API webhook.

---

BTW would be happy to know more about real-world use cases you are thinking about for this idea. Maybe we can find out more elegant solutions.

**Reply by maritaria** [reply]
This explanation would be a great addition to the documentation on @nuxt/kit, also the explanation of context I've not read anywhere else before but really helps make the concept of SSR and client code (how nuxt approaches it) click for me.
